File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-0.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    good = 0
    
    goodNodes(root) {
        this.foo(root, root.val)
        return this.good
    }

    foo(root, max) {
        if(!root) return
        if(root.val >= max) {
            console.log(max)
            max = root.val
            this.good++
        }
        this.foo(root.left, max)
        this.foo(root.right, max)
    }
}

File: Data Structures & Algorithms/binary-tree-diameter/submission-0.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int diameterOfBinaryTree(TreeNode *root) {
  int diameter = 0;
  function<int(TreeNode*)> depthAndDiameter = [&](TreeNode* node) -> int {
    if (node == nullptr) {
      return 0;
    }

    int leftDepth = depthAndDiameter(node->left);
    int rightDepth = depthAndDiameter(node->right);

    diameter = max(diameter, leftDepth + rightDepth);

    return 1 + max(leftDepth, rightDepth);
  };

  depthAndDiameter(root);
  return diameter;
}
};
